indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -0,0 +1,266 @@
+            false
+        ),
+        tuningConfig
+    );
+  }
+
+  private static Pair<Map<DataSegment, File>, List<TimelineObjectHolder<String, DataSegment>>> prepareSegments(
+      TaskToolbox toolbox,
+      String dataSource,
+      Interval interval
+  ) throws IOException, SegmentLoadingException
+  {
+    final List<DataSegment> usedSegments = toolbox
+        .getTaskActionClient()
+        .submit(new SegmentListUsedAction(dataSource, interval, null));
+    final Map<DataSegment, File> segmentFileMap = toolbox.fetchSegments(usedSegments);
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = CompactionTaskUtils.toTimelineSegments(
+        usedSegments,
+        interval
+    );
+    return Pair.of(segmentFileMap, timelineSegments);
+  }
+
+  private static DataSchema createDataSchema(
+      String dataSource,
+      Interval interval,
+      IndexIO indexIO,
+      ObjectMapper jsonMapper,
+      List<TimelineObjectHolder<String, DataSegment>> timelineSegments,
+      Map<DataSegment, File> segmentFileMap
+  )
+      throws IOException, SegmentLoadingException
+  {
+    // find metadata for interval
+    final List<QueryableIndex> segments = CompactionTaskUtils.loadSegments(timelineSegments, segmentFileMap, indexIO);
+    final Granularity expectedGranularity = segments.get(0).getMetadata().getQueryGranularity();
+    final boolean expectedRollup = segments.get(0).getMetadata().isRollup();
+
+    // check metadata
+    Preconditions.checkState(
+        segments.stream().allMatch(index -> index.getMetadata().getQueryGranularity().equals(expectedGranularity))
+    );
+    Preconditions.checkState(
+        segments.stream().allMatch(index -> index.getMetadata().isRollup() == expectedRollup)
+    );
+
+    // find merged aggregators
+    final List<AggregatorFactory[]> aggregatorFactories = segments.stream()
+                                                                  .map(index -> index.getMetadata().getAggregators())
+                                                                  .collect(Collectors.toList());
+    final AggregatorFactory[] mergedAggregators = AggregatorFactory.mergeAggregators(aggregatorFactories);
+    // conver to combining aggregators
+    final AggregatorFactory[] combiningAggregators;
+    if (mergedAggregators == null) {
+      combiningAggregators = null;
+    } else {
+      final List<AggregatorFactory> combiningAggregatorList = Arrays.stream(mergedAggregators)
+                                                                    .map(AggregatorFactory::getCombiningFactory)
+                                                                    .collect(Collectors.toList());
+      combiningAggregators = combiningAggregatorList.toArray(new AggregatorFactory[combiningAggregatorList.size()]);
+    }
+
+    // find granularity spec
+    final GranularitySpec granularitySpec = new UniformGranularitySpec(
```

Comment:
I think using ArbitraryGranularitySpec would be simpler. It's designed to index specific intervals.

Reply:
Changed.